src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.cs

```diff
@@ -91,18 +92,15 @@ internal void Serialize(Stream serializationStream, object graph, bool check)
 
         internal static TypeInformation GetTypeInformation(Type type)
         {
-            lock (s_typeNameCache)
+            TypeInformation typeInformation;
+            if (!s_typeNameCache.TryGetValue(type, out typeInformation))
             {
-                TypeInformation typeInformation;
-                if (!s_typeNameCache.TryGetValue(type, out typeInformation))
-                {
-                    bool hasTypeForwardedFrom;
-                    string assemblyName = FormatterServices.GetClrAssemblyName(type, out hasTypeForwardedFrom);
-                    typeInformation = new TypeInformation(FormatterServices.GetClrTypeFullName(type), assemblyName, hasTypeForwardedFrom);
-                    s_typeNameCache.Add(type, typeInformation);
-                }
-                return typeInformation;
+                bool hasTypeForwardedFrom;
+                string assemblyName = FormatterServices.GetClrAssemblyName(type, out hasTypeForwardedFrom);
+                typeInformation = new TypeInformation(FormatterServices.GetClrTypeFullName(type), assemblyName, hasTypeForwardedFrom);
+                s_typeNameCache.TryAdd(type, typeInformation);
             }
+            return typeInformation;
```

Comment:
This could be written more succinctly using the GetOrAdd method, e.g.:
```C#
internal static TypeInformation GetTypeInformation(Type type) =>
    s_typeNameCache.GetOrAdd(type, t =>
    {
        string assemblyName = FormatterServices.GetClrAssemblyName(t, out bool hasTypeForwardedFrom);
        return new TypeInformation(FormatterServices.GetClrTypeFullName(t), assemblyName, hasTypeForwardedFrom);
    });
```
In addition to being smaller, this also guarantees that the same instance will be returned for a specific type, even if multiple threads race to initialize it.  That's also true in the locking version.  But in the PR version, it's not true because two threads could each create their own typeInformation and return the one they created even if the call to TryAdd fails.